// mediabunny/dist/modules/src/index.js
/*!
 * Copyright (c) 2025-present, Vanilagy and contributors
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
/// <reference types="dom-mediacapture-transform" preserve="true" />
/// <reference types="dom-webcodecs" preserve="true" />
export { Output, ALL_TRACK_TYPES, } from './output.js';
export { OutputFormat, IsobmffOutputFormat, Mp4OutputFormat, MovOutputFormat, MkvOutputFormat, WebMOutputFormat, Mp3OutputFormat, WavOutputFormat, OggOutputFormat, AdtsOutputFormat, } from './output-format.js';
export { MediaSource, VideoSource, EncodedVideoPacketSource, VideoSampleSource, CanvasSource, MediaStreamVideoTrackSource, AudioSource, EncodedAudioPacketSource, AudioSampleSource, AudioBufferSource, MediaStreamAudioTrackSource, SubtitleSource, TextSubtitleSource, } from './media-source.js';
export { VIDEO_CODECS, PCM_AUDIO_CODECS, NON_PCM_AUDIO_CODECS, AUDIO_CODECS, SUBTITLE_CODECS, Quality, QUALITY_VERY_LOW, QUALITY_LOW, QUALITY_MEDIUM, QUALITY_HIGH, QUALITY_VERY_HIGH, } from './codec.js';
export { canEncode, canEncodeVideo, canEncodeAudio, canEncodeSubtitles, getEncodableCodecs, getEncodableVideoCodecs, getEncodableAudioCodecs, getEncodableSubtitleCodecs, getFirstEncodableVideoCodec, getFirstEncodableAudioCodec, getFirstEncodableSubtitleCodec, } from './encode.js';
export { Target, BufferTarget, StreamTarget } from './target.js';
export { Source, BufferSource, StreamSource, BlobSource, UrlSource, } from './source.js';
export { InputFormat, IsobmffInputFormat, Mp4InputFormat, QuickTimeInputFormat, MatroskaInputFormat, WebMInputFormat, Mp3InputFormat, WaveInputFormat, OggInputFormat, ALL_FORMATS, MP4, QTFF, MATROSKA, WEBM, MP3, WAVE, OGG, } from './input-format.js';
export { Input } from './input.js';
export { InputTrack, InputVideoTrack, InputAudioTrack } from './input-track.js';
export { EncodedPacket } from './packet.js';
export { VideoSample, AudioSample, } from './sample.js';
export { EncodedPacketSink, BaseMediaSampleSink, VideoSampleSink, CanvasSink, AudioSampleSink, AudioBufferSink, } from './media-sink.js';
export { Conversion } from './conversion.js';
export { CustomVideoDecoder, CustomAudioDecoder, CustomVideoEncoder, CustomAudioEncoder, registerDecoder, registerEncoder, } from './custom-coder.js';
// 🐡🦔
